### ISSUE TITLE
Launch with PoA validator set which can be switchted to NoPoS

### ISSUE THREAD

**Issue by bedeho** [Joystream/joystream#3802]
# Background
As part of `Carthage` Jsgensis will operate `PoA` validators as part of the first stage of roleout, and then transition to `NoPoS`, as described here https://github.com/Joystream/joystream/issues/3792.

# Proposal
Update runtime so as to allow it to start in PoA mode, and then through a `Root` origin extrinsic transition to `NoPoS`. Initial validator set is set in genesis config.

## Resources
* https://www.gautamdhameja.com/resilient-poa-network-substrate/
* https://github.com/gautamdhameja/substrate-poa
* https://docs.substrate.io/tutorials/v3/private-network/

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202132419573087/1202346732309222) by [Unito](https://www.unito.io)

**Comment by mnaamani** [comment]
Additional Resource: https://substrate.stackexchange.com/questions/996/how-to-upgrade-consensus-engine-of-a-live-chain-from-poa-to-pos/1038#1038

**Comment by dobertRowneySr** [comment]
@mnaamani 
I am not 100% sure about the role of these pallets in our runtime. Can you please provide a small description about those market with a "?".
I have used a reference substrate node template https://github.com/substrate-developer-hub/substrate-node-template
## Frame pallets are used in JS runtime and not in the node template
- Aura, tied to block validation by authority
## Frame pallets are used in JS runtime and not in the node template
- Utility substrate_utility, various and unrelated to block production/finalization
- ? BabeNPoS block production
- ? Authorship
- ? ElectionProviderMultiPhase, is this used in council election
- ? Staking: handles validator staking
- Session: handle a period of time that has a constant set of validators, it particularly takes care of maintaining the validator set
- ? Historical
- ? AuthorityDiscovery: retrieve the current set of authorities,
- ImOnline: keeps track of the active validators,
-  Offences: Tracks reported offences (i.e. slash stacked amount from validators)
- ? BagsList
- Vesting, related to the vesting of funds
- Multisig, multisig

My plan is to 
- [ ] remove all of those related to PoS from `runtime/src/lib.rs` and keep Aura only for the PoA version
- [ ] adjust chainspec if necessary so that Aura can have its validator set at genesis
- [ ] figure out a way to transfer validator set during migration 
- [ ] for the Npos I guess we should use the current runtime pallets configuration

**Comment by dobertRowneySr** [comment]
So far I know that having: pallets
- Aura
- Grandpa

is sufficient for PoA consensus, the real problem is whenever I try to remove a pallet that is relevant for PoA it breaks the runtime

**Comment by bedeho** [comment]
# Meeting Notes

## TODO

- Ignazio: Update genesis build code: initial Jsgenesis validators must stake using real existing funds and keys.
- Ignazio: Figure out how to update validator set with Sudo, currently we don't know.
- Mokhtar: review Polkadot launch code
- Mokhtar+Bedeho: Determine whether there are any problems that can arise from
   - staking pallet autonomously slashes people for not producing blocks.
   - imonline slashing
   - noone showing up
- Everyone: once staking network is unblocked, review all validation and nominatino pallets to developer deeper understanding


## Requirements

### PoA

- staking: yes, because this sets stage for new validators automatically, and no unfair entry.
- rewards: no
- tip: no
- points: no
- update validators by sudo: yes
- block producer selection: ideally BABE
- block other extrinsics not needed for any of this.

### NoPoS

- Mode change has to be possible at any arbitrary block <= added after meeting
- Has to work even if no third party shows up to validate or stake
- Rewards, tips, points: turned on

## Current Proposal

### PoA

- filter all not in {staking,sudo, utility, system?}
- include all pallets for NoPoS 
- reward: 0
- tip: 0
- points: 0
- block producer selection: BABE, configured with epochchangetrigger=sameforever
- session pallet: ?
- grandpa pallet: ?

### Migration
- Nothing

### NoPoS

- Allow all pallets
- Include all pallets
- Rewards, tips, points: allowed
- Block producer selection: BABE, configured with epochchangetrigger=external trigger

**Comment by dobertRowneySr** [comment]
### Proposals for PoA
#### PoA Genesis

1.  set authorities in session using chainspec: [we are already doing that](https://github.com/Joystream/joystream/blob/5735ebf72458bbf06a3a4c9a9f446e73864ff8b5/bin/node/src/chain_spec/mod.rs#L244)
2.  set forceEra in genesis using `ForceNone` to block era production, this effectively enables
    PoA as `on_session` is none. This effectively blocks staking pallet' s validators election.

#### Replacing keys during PoA
-   ~Any authority set update now can be done directly on the session pallet via `set_keys` call~


#### Switch to NPoS
-   using `pallet_staking::force_new_era(root_origin)` should re-enable normal election cycle hence enabling normal DPoS/NPoS.

#### Reward (draft)
- No era triggered, no era payouts to be awarded

#### Consequences
- we can keep the `pallet_babe::EpochChangeTrigger = External`
- we can keep the staking pallet active

#### References
- Polkadot configuration https://github.com/paritytech/polkadot/blob/v0.8.2/runtime/polkadot/src/lib.rs
- https://substrate.stackexchange.com/questions/4899/babe-poa-to-npos-migration
